illumos-utils/src/zfs.rs

```diff
@@ -421,8 +486,12 @@ impl Zfs {
         cmd.arg(DatasetProperties::ZFS_GET_PROPS);
         cmd.args(datasets);
 
-        let output = execute(cmd).with_context(|| {
-            format!("Failed to get dataset properties for {datasets:?}")
+        // We are intentionally ignoring the output status of this command.
+        //
+        // If one or more dataset doesn't exist, we can still read stdout to
+        // see about the ones that do exist.
+        let output = cmd.output().map_err(|err| {
```

Comment:
Here's an easy example to see what's happening:

```
$ zfs get -o name,property,value used rpool does_not_exist
cannot open 'does_not_exist': dataset does not exist
NAME   PROPERTY  VALUE
rpool  used      256G
```

"Cannot open 'does_not_exist'" goes to stderr, the rest goes to stdout, and can be parsed.
Since we want to use this API to get as much info as possible about datasets which may or may not exist, it's critical to keep going, even if one or more datasets are missing.

Reply:
Yeah, I don't know how to proceed here. The exit code is "1", so there isn't much information to glean from this.

I looked through the `zfs get` flags and didn't see anything that allows this omission. I think my **preference** with these tools would be to rely on other mechanisms to detect problems with the datasets, and leave the listing as a best-effort thing.

Namely: If a dataset that should exist actually doesn't, for any reason, we'll try to re-create it later as a part of blueprint execution. Then we'll see errors at "ensure" time, rather than "listing" time. 

> I guess we could do something like "on error, try to parse stderr, and if it's only dataset does not exist logs then proceed"?

This is possible - to be clear, would this proposal be:

- Call `zfs get` on several datasets (like this PR is currently doing)
- If we see the command fail, parse stderr
- Only treat it as "OK" if stderr contains the lines `cannot open 'XXX': dataset does not exist` for one of our datasets?

Small quirk here -- if this failed for any datasets we're trying to query (e.g., maybe a zpool is dying?) it would stop inventory collection for the entire sled. So at the end of the day, I *think* we need to proceed, parsing what we can, from the set of all datasets we can observe.

> Or maybe we could call zfs get without listing any datasets at all, then filter the returned list ourselves to match which?

I also considered this, and would be somewhat okay with this approach. My only beef is that it would need to be fully recursive to be **able** to see everything, so we'd be e.g. listing all crucible datasets every time -- and that means there could be a *lot* of data here.

For example, here are the results on a "non-pathological system" -- dogfood:

```
pilot host exec -c "zfs get -o name,value name | wc -l" {7..25}
 7  BRM27230045        ok: 108
 8  BRM44220011        ok: 4364
 9  BRM44220005        ok: 4190
10  BRM42220009        ok: 263
11  BRM42220006        ok: 4298
12  BRM42220057        ok: 4463
13  BRM42220018        ok: 21
14  BRM42220051        ok: 4281
16  BRM42220014        ok: 4397
17  BRM42220017        ok: 4286
21  BRM42220031        ok: 4315
23  BRM42220016        ok: 4379
```

Reply:
> Would it be possible/reasonable here to specify a parent dataset that _does_ exist and use `-r` to get _all_ its children? Then we shouldn't expect any errors.

I think this bumps into the problem I mentioned about recursive children - we *do* want to know about e.g. crucible, but using `-r` will list all regions, and all properties on all regions, which is in the ballpark of 10,000 - 100,000 lines of output.